samples: bluetooth: add L2CAP CoC samples

ritesh006 opened the pull request:
Adds a "Creating a simple L2CAP server" section to the L2CAP API docs with
a minimal skeleton example and step-by-step instructions.

Also fixes minor typos (enabled/latter).

Fixes: #96494

ritesh006 commented:
Ready for Review

HaavardRei commented on doc/connectivity/bluetooth/api/l2cap.rst:
```diff
@@ -65,9 +65,101 @@ listen to, the required security level ``sec_level``, and the callback
+
+Notes
+~~~~~
+* For multiple concurrent connections allocate channel instances dynamically (k_malloc or pool) instead of reusing a single ``my_chan``.
```
You could also allocate channel instances based on the `CONFIG_BT_MAX_CONN` parameter, similar to: https://github.com/zephyrproject-rtos/zephyr/blob/0ab40913c2de53ee504327761b37c367cb544944/doc/connectivity/bluetooth/api/l2cap.rst?plain=1#L17
and 
https://github.com/zephyrproject-rtos/zephyr/blob/0ab40913c2de53ee504327761b37c367cb544944/doc/connectivity/bluetooth/api/l2cap.rst?plain=1#L29-L35

cvinayak reviewed (dismissed):
Please add a sample under `samples/bluetooth` instead which is CI tested and will not go stale. You can refer to the sample here. And hopefully be able to have directly reference to the code in the sample in this document.

HaavardRei commented:
> Please add a sample under `samples/bluetooth` instead which is CI tested and will not go stale. You can refer to the sample here. And hopefully be able to have directly reference to the code in the sample in this document.

There is already various bsim test covering this functionality, for instance: https://github.com/zephyrproject-rtos/zephyr/blob/0ab40913c2de53ee504327761b37c367cb544944/tests/bsim/bluetooth/host/l2cap/send_on_connect/src/main_l2cap_send_on_connect.c#L128-L141 I don't think adding a sample for the sole purpose of testing this in CI is necessary, though I agree that the functionality should be demonstrated in a sample.

cvinayak commented:
> What could be a compromise here (keeping doc part and implementing a sample with covering it in CI) is using literalinclude doxygen directive, for example: 

Yes, this is what I want, have the documentation but reference to actual code blocks from samples that are being built in CI (not necessarily CI tested, but that will be great too) and which do not go stale over time. A reader can copy these code blocks as instructed by the documentation.

ritesh006 commented:
Thank you all! I added a real example at samples/bluetooth/l2cap_server_simple and updated the docs to show the code with literalinclude. The sample shows how to use fixed space for each connection (it uses CONFIG_BT_MAX_CONN and bt_conn_index(conn)). The docs example is now copied from this file.

I am new here and still learning, so please let me know if I missed something or can do better. I’m trying hard to help and get things right.

PavelVPV commented:
@ritesh006 ,

I think what needs to be done to complete this task is:
- another sample that will connect the channels
- demonstrate data transmission
	- I think it is Ok for sample if it sends some random printable data over channels with some delay, but have no strong opinion what and how exactly samples should send data.
- a short readme for both the samples (see any other sample)

I can recommend you to look at [l2cap ecred generic](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bsim/bluetooth/host/l2cap/general/src/main_l2cap_ecred.c#L38) babblesim test.
This can look a bit complicated, but can on the other hand help to get the idea what API needs to be used for communication without digging into babblesim API.

The code in both samples should not be overcomplicated (and your current sample in the PR is actually concise). You can look at, for example, [peripheral](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/peripheral/src/main.c) and [central](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/central/src/main.c) samples to see how much code you need for both sample.

Regarding documentation, I suggest to use `literalinclude` instead of [this](https://github.com/zephyrproject-rtos/zephyr/blob/02920b00d5f8c10fff701d7a243adc8a78e31e3d/doc/connectivity/bluetooth/api/l2cap.rst?plain=1#L80-L153) text so that documentation is automatically corrected if something (e.g. API) changes.

kartben reviewed (changes requested):
Missing license and copyright headers also need to be added

PavelVPV reviewed (commented):
I haven't reviewed READMEs yet. Will do this after @kartben comments are resolved there.

PavelVPV commented on samples/bluetooth/l2cap_coc_acceptor/README.rst:
```diff
@@ -0,0 +1,89 @@
+.. _bluetooth_l2cap_coc_acceptor:
+
+L2CAP Connection Oriented Channels (Acceptor)
+##############################################
```
```suggestion
L2CAP Connection Oriented Channels (Acceptor)
#############################################
```

PavelVPV commented on doc/connectivity/bluetooth/api/l2cap.rst:
```diff
@@ -8,6 +8,40 @@ configuration option: :kconfig:option:`CONFIG_BT_L2CAP_DYNAMIC_CHANNEL`. These c
+
+.. literalinclude:: ../../../../samples/bluetooth/l2cap_coc_acceptor/src/main.c
+   :language: c
+   :lines: 17-48
```
Lets use references here to comments like e.g. usb documentation does instead of code lines as they can change:
https://github.com/zephyrproject-rtos/zephyr/blob/2d72d86ba2d561d76243879dd0b29edbf9ab7608/samples/subsys/usb/common/sample_usbd_init.c#L24-L33
https://github.com/zephyrproject-rtos/zephyr/blob/2d72d86ba2d561d76243879dd0b29edbf9ab7608/doc/connectivity/usb/device_next/usb_device.rst?plain=1#L69-L73

jhedberg commented:
@ritesh006 please split this into three commits:
1. Changes under `doc/`
2. initiator (should this be called client instead?)
3. acceptor (should this be called server instead?)

ritesh006 committed:
doc: add L2CAP CoC documentation example

Add a step-by-step example for creating an L2CAP CoC server and update the
documentation to reference the new samples.

Signed-off-by: Ritesh Kudkelwar <ritesh.kumar0793@gmail.com>

ritesh006 committed:
samples: bluetooth: add L2CAP CoC client sample

Add a client sample that connects to an L2CAP CoC server and demonstrates
basic data transmission.

Signed-off-by: Ritesh Kudkelwar <ritesh.kumar0793@gmail.com>

ritesh006 committed:
samples: bluetooth: add L2CAP CoC server sample

Add a simple L2CAP Connection-oriented Channel (CoC) server sample that
accepts incoming connections and exchanges data.

Signed-off-by: Ritesh Kudkelwar <ritesh.kumar0793@gmail.com>

ritesh006 commented on samples/bluetooth/l2cap_coc_acceptor/src/main.c:
Good point, thanks for catching that.
Renamed fixed_chan to l2cap_chans to avoid confusion with fixed L2CAP channels, since the sample uses dynamic CoC channels with pre-allocated storage.

ritesh006 committed:
samples: bluetooth: clarify dynamic L2CAP channel naming

Rename fixed_chan to l2cap_chans to avoid confusion with fixed channels.

Signed-off-by: Ritesh Kudkelwar <ritesh.kumar0793@gmail.com>

ritesh006 commented:
@cvinayak could you please review this
